pandas/core/missing.py

```diff
@@ -160,35 +160,41 @@ def _interp_limit(invalid, fw_limit, bw_limit):
     start_nans = set(range(ys.first_valid_index()))
     end_nans = set(range(1 + ys.last_valid_index(), len(valid)))
 
-    # This is a list of the indexes in the series whose yvalue is currently
-    # NaN, but whose interpolated yvalue will be overwritten with NaN after
-    # computing the interpolation. For each index in this list, one of these
-    # conditions is true of the corresponding NaN in the yvalues:
+    # violate_limit is a list of the indexes in the series whose yvalue is
+    # currently NaN, and should still be NaN after the interpolation.
+    # Specifically:
     #
-    # a) It is one of a chain of NaNs at the beginning of the series, and
-    #    either limit is not specified or limit_direction is 'forward'.
-    # b) It is one of a chain of NaNs at the end of the series, and limit is
-    #    specified and limit_direction is 'backward' or 'both'.
-    # c) Limit is nonzero and it is further than limit from the nearest non-NaN
-    #    value (with respect to the limit_direction setting).
+    # If limit_direction='forward' or None then the list will contain NaNs at
+    # the beginning of the series, and NaNs that are more than 'limit' away
+    # from the prior non-NaN.
     #
-    # The default behavior is to fill forward with no limit, ignoring NaNs at
-    # the beginning (see issues #9218 and #10420)
-    violate_limit = sorted(start_nans)
-
-    if limit is not None:
-        if not is_integer(limit):
-            raise ValueError('Limit must be an integer')
-        if limit < 1:
-            raise ValueError('Limit must be greater than 0')
-        if limit_direction == 'forward':
-            violate_limit = sorted(start_nans | set(_interp_limit(invalid,
-                                                                  limit, 0)))
-        if limit_direction == 'backward':
-            violate_limit = sorted(end_nans | set(_interp_limit(invalid, 0,
-                                                                limit)))
-        if limit_direction == 'both':
-            violate_limit = sorted(_interp_limit(invalid, limit, limit))
+    # If limit_direction='backward' then the list will contain NaNs at
+    # the end of the series, and NaNs that are more than 'limit' away
+    # from the subsequent non-NaN.
+    #
+    # If limit_direction='both' then the list will contain NaNs that
+    # are more than 'limit' away from any non-NaN.
+    #
+    # If limit=None, then use default behavior of filling an unlimited number
+    # of NaNs in the direction specified by limit_direction
+
+    # default limit is unlimited GH #16282
+    if limit is None:
+        limit = len(xvalues)
+    elif not is_integer(limit):
+        raise ValueError('Limit must be an integer')
+    elif limit < 1:
+        raise ValueError('Limit must be greater than 0')
+
+    # each possible limit_direction
+    if limit_direction == 'forward':
+        violate_limit = sorted(start_nans |
```

Comment:
@TomAugspurger not sure if this is where the perf issue is (maybe because not this IS hit in all code).

these can be done via Index set ops

IOW

```
start_nans = Index(range(.....))

violate_limit = start_nans.union(Index(_interp_limit(invalid, limit, 0))).sort_values()
```
for example